### PR TITLE
[Fix #3438] Add Style/FormatStringToken cop

### LINKED ISSUE #3438: Cop idea: Enforce preferred style for format string sequences
You should use one of these styles consistently:

``` ruby
# Positional format sequences with field type characters
"Hi, my name is %s and I am %d years old. %0.1f to be exact." % ["John", 22, 22.45]

# Named format sequences without field type characters
"Hi, my name is %{name} and I am %{age} years old. %{precise_age} to be exact." % { name: "John", age: 22, precise_age: 22.45 }

# Named format sequences with field type characters
"Hi, my name is %<name>s and I am %<age>d years old. %<age>0.1f to be exact." % { name: "John", age:  22.45 }
```

I prefer the third option because you get to name your variables and also get some type hinting and coercion.

### PR DESCRIPTION
**Pull request by backus**
Enforces using one of many format string token styles:

 * Unnamed:  `format('%s', 'Hello')`
 * Named:    `format('%<greeting>s', greeting: 'Hello')`
 * Template: `format('%{greeting}', greeting: 'Hello')`

I haven't corrected the rubocop codebase because I'm not actually sure what the preferred style is. I think it should be `%<named>s` style tokens for a few reasons:

1. You get a bit of type enforcement when you use them. Ruby will raise an error if you write `%<named>d' and give it a string

2. It is more clear when you actually end up passing the values to the format method. Positional arguments get confusing to track through after one or two and it is unnecessarily tricky to figure out how to reuse unnamed positional tokens.

Thoughts? I can move this to a ruby style guide issue but I actually think it is interesting to discuss here since the rubocop codebase itself is split between different styles.

-----------------

Before submitting the PR make sure the following are checked:

* [x] Wrote [good commit messages][1].
* [x] Commit message starts with `[Fix #issue-number]` (if the related issue exists).
* [ ] Used the same coding conventions as the rest of the project.
* [x] Feature branch is up-to-date with `master` (if not - rebase it).
* [ ] Squashed related commits together.
* [ ] Added tests.
* [ ] Added an entry to the [Changelog](../blob/master/CHANGELOG.md) if the new code introduces user-observable changes. See [changelog entry format](../blob/master/CONTRIBUTING.md#changelog-entry-format).
* [ ] All tests are passing.
* [ ] The new code doesn't generate RuboCop offenses.
* [x] The PR relates to *only* one subject with a clear title
  and description in grammatically correct, complete sentences.
* [ ] Updated cop documentation with `rake generate_cops_documentation` (required only when you've added a new cop or changed the configuration/documentation of an existing cop).

[1]: http://chris.beams.io/posts/git-commit/

### CODE REVIEW

**Comment by Drenmi** [comment]
Despite being a bit clunky, I also prefer the "named" style, for the reasons you highlighted.

**Comment by backus** [comment]
btw @Drenmi I was *dying* for a `RuboCop::AST::StrNode` while working on this. A lot of the code for this cop belongs in that object eventually

**Comment by Drenmi** [comment]
@backus: Duly noted. It is definitely towards the top of my list, but because of how `parser` handles interpolated strings, it might not be as straight forward as some other, lower hanging, fruit. 🙂

**Comment by bbatsov** [comment]
> Despite being a bit clunky, I also prefer the "named" style, for the reasons you highlighted.

I keep learning new stuff still. :-) Even if I knew about this at some point I managed to totally forget about it. :D I've favoured the template style in recent years, but I agree that the `named` style has advantages over it. I also think we should have something in the Ruby style guide on the subject.

File: CHANGELOG.md [hunk 1]
```diff
@@ -2,6 +2,10 @@
 
 ## master (unreleased)
 
+### New features
+
+* [#3438](https://github.com/bbatsov/rubocop/pull/4237): Add new `Style/FormatStringToken` cop. ([@backus][])
```
Comment (bbatsov) [inline CHANGELOG.md]: 4237 -> 3438

**Comment by backus** [comment]
Here is the style guide PR https://github.com/bbatsov/ruby-style-guide/pull/650

In my original description I talked about this as also enforcing `%<named>s` over `%s` but I think that is wrong now. There are some cases where unnamed tokens are fine.

**Comment by bbatsov** [comment]
> In my original description I talked about this as also enforcing %<named>s over %s but I think that is wrong now. There are some cases where unnamed tokens are fine.

Well, there are no cases where more clarity would hurt, right? :-) But I understand what you mean - this might be cryptic but many people are kinda used to it.

**Comment by bbatsov** [comment]
@backus ping :-)

**Comment by backus** [comment]
I'll try to have the cop updated this weekend. You can temporarily close it if I don't have it done by then

**Comment by backus** [comment]
> Well, there are no cases where more clarity would hurt, right? :-) But I understand what you mean - this might be cryptic but many people are kinda used to it.

Thinking about it a bit more, I think the best solution is another cop that says to used named format tokens if you have more than 1 or 2 positional tokens.

**Comment by bbatsov** [comment]
> Thinking about it a bit more, I think the best solution is another cop that says to used named format tokens if you have more than 1 or 2 positional tokens.

I agree. Guess this number should be configurable.

**Pull request merged**